### PR TITLE
Dcadams/course tiles from db

### PR DESCRIPTION
**Pull request by dcadams**
Provides option for having tiles from db table.

A new model, TileConfiguration, in the branding app provides the
storage for course tiles driven from the db.

A django admin interface is provided for population and editing
of said model.

Note, the deployment of this PR to prod must include the manual
population of the table through the django admin interface.
##
 
Provides a setting for display of course tiles on home page.

A new setting, DISPLAY_COURSE_TILES, acts as a boolean for the display of
course tiles on the home page.

In our system, this will be True for Lagunita and False for Suclass.

### CODE REVIEW

**Comment by dcadams** [comment]
@stvstnfrd Here's the changes for course tiles as discussed.

**Comment by dcadams** [comment]
@kluo

File: lms/envs/aws.py [hunk 1]
```diff
@@ -243,6 +243,7 @@
                             service_variant=SERVICE_VARIANT)
 
 COURSE_LISTINGS = ENV_TOKENS.get('COURSE_LISTINGS', {})
```
Comment (kluo) [inline lms/envs/aws.py]: Do we need to get rid of this setting everywhere?

File: lms/djangoapps/branding/models.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
+"""
+Model for branding
+
+This was implemented initially so that the list of course tiles could be
+stored in this model.
+"""
+
+#from django.contrib import admin
```
Comment (kluo) [inline lms/djangoapps/branding/models.py]: Delete

File: lms/djangoapps/branding/models.py [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+from django.db import models
+
+from config_models.models import ConfigurationModel
+from xmodule_django.models import CourseKeyField
+
+
+class TileConfiguration(ConfigurationModel):
+    """
+        Stores a list of tiles presented on the front page.
```
Comment (kluo) [inline lms/djangoapps/branding/models.py]: Excess indentation

File: lms/djangoapps/branding/migrations/0001_initial.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
+# -*- coding: utf-8 -*-
```
Comment (stvstnfrd) [inline lms/djangoapps/branding/migrations/0001_initial.py]: In general, I'm hesitant to add our own migrations to an existing djangoapp. This isn't an issue yet, but could be if edX ever adds their own model here.

This is how we run into issues w/ migrations and need to workaround with `--delete-ghost-migrations`, `--skip-ghost-migrations`, and `--merge` because DirectAccess piggy-backs on an existing model.

Before Jason left, he and I actually drafted a hypothetical plan for moving DirectAccess _out_ of `StudentProfile` and into its own model.

Is there a way for us to compartmentalize this? I'm not black-and-white on this, so let me know your thoughts.

File: lms/djangoapps/branding/models.py [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+    """
+    site = models.CharField(max_length=32, default='default', blank=False)
+    course_id = CourseKeyField(max_length=128, db_index=True)
```
Comment (stvstnfrd) [inline lms/djangoapps/branding/models.py]: Is there a constant value somewhere we can reference for the `max_length`? 

We probably have this magic number littered throughout the code base, but it would be nice if we could avoid doing so...

File: lms/djangoapps/branding/models.py [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+
+    def __unicode__(self):
+        return "{0} {1} {2}".format(self.site, self.course_id, self.enabled)
```
Comment (stvstnfrd) [inline lms/djangoapps/branding/models.py]: Does this need to be:
```python
return u"..."
```

like this?

File: lms/djangoapps/branding/__init__.py [hunk 7]
```diff
@@ -10,6 +11,11 @@ def get_visible_courses():
     """
     Return the set of CourseDescriptors that should be visible in this branded instance
     """
+
+    # In the event we don't want any course tiles displayed
+    if not (hasattr(settings, 'DISPLAY_COURSE_TILES') and settings.DISPLAY_COURSE_TILES):
```
Comment (stvstnfrd) [inline lms/djangoapps/branding/__init__.py]: Can this be simplified?
```python
if not getattr(settings, 'DISPLAY_COURSE_TILES', False):
    return []
```

**Comment by stvstnfrd** [comment]
How is this query cached?

Since we'll be running it on every home page load and this list seldom changes, this value should be cached for a considerable amount of time.

Disclaimer: I'm naive to our memcache configuration.

**Comment by stvstnfrd** [comment]
Tangentially, would this be a good candidate to "package and release" to edx-code?

File: lms/djangoapps/branding/migrations/0001_initial.py (hunk 4 above)
Comment (dcadams) [inline lms/djangoapps/branding/migrations/0001_initial.py]: In general adding our own migrations to an an existing app would not be a problem on production. The migrations are sequenced <and> named such that there should never be a conflict. The ghost migration issue should only affect dev environments when one goes back and forth between origin and upstream branches and I'm not at all sympathetic to doing anything different there. That having been said, I did add an initial migration to an upstream app, which I think would cause a file level conflict, so I've moved it to its own app. Thanks for noticing.

File: lms/djangoapps/branding/models.py (hunk 5 above)
Comment (dcadams) [inline lms/djangoapps/branding/models.py]: It littered and duly inconsistent. I think fixing that problem is out of scope for this PR.

File: lms/envs/aws.py (hunk 1 above)
Comment (dcadams) [inline lms/envs/aws.py]: I don't think so; it can still be used.

**Comment by dcadams** [comment]
There's actually very little mention of memache in the codebase. When I looked at memcache a few years back, in another context, I seem to recall it provided a layer outside of the app. It could be that we have memcache running on prod and the query will be cached but I don't know. In any event, in the scheme of things, I'm not worried about this simple single table query.

File: lms/djangoapps/branding/models.py (hunk 5 above)
Comment (stvstnfrd) [inline lms/djangoapps/branding/models.py]: Yeah, I'm not worried about making this consistent everywhere (now). But if there _is_ a constant _already_ defined, we should use it. I didn't know if there was one, more curiosity than anything.

File: lms/djangoapps/branding/migrations/0001_initial.py (hunk 4 above)
Comment (stvstnfrd) [inline lms/djangoapps/branding/migrations/0001_initial.py]: Thanks for the explanation!

**Comment by dcadams** [comment]
Hopefully, everything addressed.

File: lms/djangoapps/branding/__init__.py [hunk 8]
```diff
@@ -27,6 +33,14 @@ def get_visible_courses():
 
     filtered_by_org = microsite.get_value('course_org_filter')
 
+    filtered_by_db = TileConfiguration.objects.filter(
+        enabled=True,
+    ).values('course_id').order_by('-change_date')
+
+    if filtered_by_db:
+        filtered_by_db_ids = [course['course_id'] for course in filtered_by_db]
+        filtered_by_db_keys = frozenset([SlashSeparatedCourseKey.from_deprecated_string(c) for c in filtered_by_db_ids])
```
Comment (stvstnfrd) [inline lms/djangoapps/branding/__init__.py]: This shouldn't assume that `CourseKeys` are always old-style triples. Since we're looking to enable split_mongo (and course_rereruns) soon, we'll need to make this OpaqueKeys-compliant and not refer to deprecated strings.

**Comment by stvstnfrd** [comment]
I really like how you split this out. This gives us a nice, clean,
purely additive `diff` which won't run the risk of conflicting w/
upstream.

I played around with some ideas on how to further compartmentalize our
code from the core edx-platform code and make our lives easier living on
the fork; it's outside the scope here but would like to discuss at some
point in the future.

After we address the deprecated_string note above, I'm +1 on this.

File: lms/djangoapps/branding_stanford/models.py [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+from config_models.models import ConfigurationModel
+from xmodule_django.models import CourseKeyField
+
+
+class TileConfiguration(ConfigurationModel):
+    """
+        Stores a list of tiles presented on the front page.
+    """
+    site = models.CharField(max_length=32, default='default', blank=False)
+    course_id = CourseKeyField(max_length=255, db_index=True)
```
Comment (stvstnfrd) [inline lms/djangoapps/branding_stanford/models.py]: So we don't need to specify `enabled` here?

File: lms/djangoapps/branding/__init__.py [hunk 10]
```diff
@@ -27,6 +33,14 @@ def get_visible_courses():
 
     filtered_by_org = microsite.get_value('course_org_filter')
 
+    filtered_by_db = TileConfiguration.objects.filter(
+        enabled=True,
+    ).values('course_id').order_by('-change_date')
+
+    if filtered_by_db:
+        filtered_by_db_ids = [course['course_id'] for course in filtered_by_db]
+        filtered_by_db_keys = frozenset([SlashSeparatedCourseKey.from_string(c) for c in filtered_by_db_ids])
```
Comment (stvstnfrd) [inline lms/djangoapps/branding/__init__.py]: Should this be `CourseKey.from_string` or some such? Will IDs always be in triple-slash-separated form?

File: lms/djangoapps/branding/__init__.py [hunk 11]
```diff
@@ -27,6 +33,14 @@ def get_visible_courses():
...
+        return [course for course in courses if course.id in filtered_by_db_keys]
```
Comment (stvstnfrd) [inline lms/djangoapps/branding/__init__.py]: Should this be a set? We shouldn't have duplicates, right?

**Comment by stvstnfrd** [comment]
Just some questions I didn't think to ask before; otherwise +1.

File: lms/djangoapps/branding_stanford/models.py (hunk 9 above)
Comment (dcadams) [inline lms/djangoapps/branding_stanford/models.py]: 'enabled' comes from the parent class: ConfigurationModel.

File: lms/djangoapps/branding/__init__.py (hunk 10 above)
Comment (dcadams) [inline lms/djangoapps/branding/__init__.py]: This comes from the course_id field in the new table which will be populated using Django Admin interface and is a reflection of the current secrets list. So, I think for now they will always be slash separated triples. My current understanding, maybe incorrect, is that when 'run' becomes part of uniqueness, it will still be a triple.

File: lms/djangoapps/branding/__init__.py (hunk 11 above)
Comment (dcadams) [inline lms/djangoapps/branding/__init__.py]: Well there shouldn't be dupes. I think the plan is to move the UI part into Sysadmin so doing the validations there would be more sensible. I'm inclined to see how we go with the Django Admin tool. We can easily fixup data in the db using Django Admin if necessary (insert, delete, edit, enable/disable).

**Comment by dcadams** [comment]
@stvstnfrd  just a reminder than this change includes a migration.

File: lms/djangoapps/branding/__init__.py (hunk 11 above)
Comment (stvstnfrd) [inline lms/djangoapps/branding/__init__.py]: No big deal, I'm not worried about this now. But in general, since Python provides `set` as a first class iterable, we should aim to use `set`s over `list`s wherever we can. 

A list comprehension can be converted to a set comprehension by just changing the square brackets into curly ones.

```python
return {
    course
    for course in courses
    if course.id in filtered_by_db_keys
}
```
